### PR TITLE
fix: fp16 kernels computed in f32

### PR DESCRIPTION
**Pull request by wjones127**
Key findings:

1. If we cast to f32, most platforms have SIMD instructions they can use to do the cast and the computation. This makes the f16 kernels as fast or faster than the existing f32 kernels.
2. The dynamic dispatch does not seems to have any meaningful impact on performance.

Changes

* Replace the `avx512fp16` flag with a general `fp16kernels` flag. This new flag is turned off by default, because it requires a fairly recent compiler (at least for GCC).
* If the `fp16kernels` feature is on, we now *require* a C compiler. Instead of hard-coding `clang`, we now also support GCC >= 12. (MSVC doesn't not support fp16 at all.)
* Added runtime detection of NEON (for Linux) and AVX2 (most x86_64 processors now).
* Added a specialized Cosine kernel, which reduces the overhead over the f16 -> f32 conversion.

### CODE REVIEW

**Comment by wjones127** [comment]
On MacOS, I've found that the kernels with SIMD are as fast or faster with fp16 vs f32, even with the conversion:

| kernel | f32 (ms) | fp16 (ms) | multiple |
|--------|--------|--------|--------|
| L2 | 96 | 55 | 40% faster |
| L2 (norm) | 73 | 45 | 38% faster |
| Dot | 90 | 68 | 24% faster |
| Cosine | 89 | 86 | same-ish |


For scalar:

| kernel | f32 (ms) | fp16 (ms) | multiple |
|--------|--------|--------|--------|
| L2 | 869 | 4,070 | 4.6x slower |
| L2 (norm) | 866 | 4,037 | 4.6x slower |
| Dot | 874 | 4,068 | 4.7x slower |
| Cosine | 957 | 4,173 | 4.3x slower |

I'll look at x86_64 next. FWIW, there seem to be SIMD kernels for converting f16 -> f32 even in AVX2. For example, here is SIMD on Broadwell: https://godbolt.org/z/3dY6xer1a. So I think we should be able to relax the fp16 flag so we can have these SIMD kernels on more architectures.

I'll also see if we can make the scalar fallbacks faster.

**Comment by wjones127** [comment]
Here are my results on x86_64 Linux (with AVX2, I assume):


| kernel | f32 (ms) | fp16 (ms) | multiple |
|--------|--------|--------|--------|
| L2 | 165 | 119 | 28% faster |
| L2 (norm) | 123 | 97 | 20% faster |
| Dot | 129 | 118 | 9% faster |
| Cosine | 145 | 122 | 16% faster |

**Comment by wjones127** [comment]
Note: it would be nice to get all versions except the avx512 ones in Rust. We could do that with [multiversion](https://docs.rs/multiversion/latest/multiversion/) or manually implement something like it with `target_feature(enable=...)` ([examples](https://alexheretic.github.io/posts/auto-avx2/)). However, in my attempt to do this now, I'm finding it's like 4x slower than the C version. It's hard to tell why. Maybe the way the f16 values are compiled down from `half` isn't quite understood by the compiler. For now, using the C file appears to be the best solution.

File: rust/lance-linalg/build.rs [hunk 1]
```diff
@@ -23,36 +23,59 @@ fn main() {
...
+            // Clang 12 and GCC 11 are the first versions with sapphire rapids support
+            eprintln!("Skipping Sapphire Rapids build due to error: {}", err);
+        } else {
+            // We create a special cfg so that we can detect we have in fact
+            // generated the AVX512 version of the f16 kernels.
+            println!("cargo:rustc-cfg=kernel_suppport=\"avx512\"");
+        };
+        // Build a version with AVX
+        // While GCC doesn't have support for _Float16 until GCC 12, clang
+        // has support for __fp16 going back to at least clang 6.
```
Comment (eddyxu) [inline rust/lance-linalg/build.rs:53]: is there a reason to use broadwell instead of haswell? 
Haswell is the oldest CPUs supported on AWS / GCP

Comment (wjones127) [reply rust/lance-linalg/build.rs:53]: I can switch this to haswell

File: rust/lance-linalg/src/simd/f16.c [hunk 2]
```diff
@@ -14,44 +14,74 @@
...
+  return sum;
+}
+
+float FUNC(cosine_f16)(const FP16 *x, float x_norm, const FP16 *y, uint32_t dimension) {
+  float dot = 0.0;
+  float l2_y = 0.0;
+
+  // Instead of using functions above, we combine the loop to reduce overhead
+  // of the fp16 to fp32 conversion.
+#pragma clang loop unroll(enable) interleave(enable) vectorize(enable)
```
Comment (chebbyChefNEQ) [inline rust/lance-linalg/src/simd/f16.c:79]: do we need different directives for gcc?

File: rust/lance-linalg/src/distance/cosine.rs [hunk 3]
```diff
@@ -89,7 +93,45 @@ where
...
+        #[cfg(target_arch = "aarch64")]
+        pub fn cosine_f16_neon(x: *const f16, x_norm: f32, y: *const f16, dimension: u32) -> f32;
+        #[cfg(all(kernel_suppport = "avx512", target_arch = "x86_64"))]
+        pub fn cosine_f16_avx512(x: *const f16, x_norm: f32, y: *const f16, dimension: u32) -> f32;
+        #[cfg(target_arch = "x86_64")]
+        pub fn cosine_f16_avx2(x: *const f16, x_norm: f32, y: *const f16, dimension: u32) -> f32;
+    }
+}
+
+impl Cosine for Float16Type {
```
Comment (chebbyChefNEQ) [inline rust/lance-linalg/src/distance/cosine.rs:112]: 🔥

**Review by chebbyChefNEQ — APPROVED** [review]
just one nit: can we add a few numerical tests? Ticket is okay.

File: rust/lance-linalg/src/simd/f16.c (hunk 2 above)
Comment (wjones127) [reply rust/lance-linalg/src/simd/f16.c:79]: TBH, I'm not even sure these directives are necessary. I will check they make a difference on godbolt.

**Pull request merged**